### PR TITLE
Foreach sem refactor

### PR DESCRIPTION
**Pull request by thewilsonator**
This is a large PR, please review by commit.

### CODE REVIEW

File: src/dmd/statementsem.d [hunk 1]
```diff
@@ -1670,306 +1766,242 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+        Expression ec;
+        const dim = fs.parameters.dim;
+        const loc = fs.loc;
+        /* Call:
+         *      _aApply(aggr, flde)
+         */
+        __gshared const(char)** fntab =
+        [
+         "cc", "cw", "cd",
+         "wc", "cc", "wd",
```
Comment (thewilsonator) [inline src/dmd/statementsem.d:1848]: `cc` here looks wrong, but was in the original.

File: src/dmd/statementsem.d [hunk 2]
```diff
@@ -1223,19 +1219,118 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
             }
         }
 
-        Statement s;
+        void retError()
+        {
+            sc2.pop();
+            result = new ErrorStatement();
+        }
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1226]: add newline

Comment (thewilsonator) [reply src/dmd/statementsem.d:1226]: done

Comment (UplinkCoder) [reply src/dmd/statementsem.d:1226]: There are multiple places with missing newlines.
I didn't want to point out every single one ;)

File: src/dmd/statementsem.d [hunk 3]
```diff
@@ -1223,19 +1219,119 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+
+        void rangeError()
+        {
+            fs.error("cannot infer argument types");
+            return retError();
+        }
+        void retStmt(Statement s)
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1233]: missing newline

File: src/dmd/statementsem.d [hunk 4]
```diff
@@ -1223,19 +1219,119 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+            if (!ec)
+                return null;
+            e = Expression.combine(e, ec);
+            return loopReturn(e, fs.cases, loc);
+        }
         switch (tab.ty)
         {
+        case Terror:
+            return retError();
+        default:
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1322]: default statement should do at the end of the case list.

File: src/dmd/statementsem.d [hunk 5]
```diff
@@ -1670,306 +1765,241 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+        ec = ec.expressionSemantic(sc2);
+        if (ec.op == TOK.error)
+            return null;
+        if (ec.type != Type.tint32)
+        {
+            fs.error("`opApply()` function for `%s` must return an `int`", tab.toChars());
+            return null;
+        }
+        return ec;
+    }
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1810]: missing newline

File: src/dmd/statementsem.d [hunk 6]
```diff
@@ -1670,306 +1765,241 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
-            s = new ErrorStatement();
-            break;
+        /* Call:
+         *  extern(C) int _aaApply(void*, in size_t, int delegate(void*))
+         *      _aaApply(aggr, keysize, flde)
+         *
+         *  extern(C) int _aaApply2(void*, in size_t, int delegate(void*, void*))
+         *      _aaApply2(aggr, keysize, flde)
+         */
+        __gshared FuncDeclaration* fdapply = [null, null];
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1937]: why is this __gshared?

File: src/dmd/statementsem.d [hunk 7]
```diff
@@ -1670,306 +1765,241 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
+            fexp.type = fldeTy[i];
+        }
+        exps.push(new IntegerExp(Loc.initial, keysize, Type.tsize_t));
+        exps.push(fexp);
+        ec = new VarExp(Loc.initial, fdapply[i], false);
+        ec = new CallExp(fs.loc, ec, exps);
+        ec.type = Type.tint32; // don't run semantic() on ec
+        return ec;
+    }
+    private static extern(D) Statement loopReturn(Expression e, Statements* cases, const ref Loc loc)
```
Comment (UplinkCoder) [inline src/dmd/statementsem.d:1975]: missing newline

**Review by UplinkCoder — CHANGES_REQUESTED** [review]
can __gshared be avoided?

File: src/dmd/statementsem.d (hunk 6 above)
Comment (thewilsonator) [reply src/dmd/statementsem.d:1937]: No idea. was like that before. I think this is supposed to be a lazily computed constant. I'll do another PR to change e.g. https://github.com/dlang/dmd/pull/12298#discussion_r599362148 I'll have another look at this to see if there's a better way to do this.

**Review by Imperatorn — COMMENTED** [review]
Any pr that reduces the number of gotos by a certain % show be automatically approved 😉

File: src/dmd/statementsem.d (hunk 4 above)
Comment (thewilsonator) [reply src/dmd/statementsem.d]: > default statement should go at the end of the case list.

done

**Comment by thewilsonator** [comment]
getting rid of the `goto`s is nice but its the partitioning of thousand line functions that makes it much easier to understand for me.

**Comment by thewilsonator** [comment]
> can __gshared be avoided?

maybe, not in this PR.

**Comment by UplinkCoder** [comment]
Alright ...

**Review by UplinkCoder — APPROVED** [review]

**Comment by WalterBright** [comment]
@thewilsonator @uplinkcoder @RazvanN7 these giant PRs need to stop. Do them as smaller PRs.

File: src/dmd/statementsem.d [hunk 8]
```diff
@@ -1825,6 +1811,30 @@ else
         result = s;
     }
 
+    private static extern(D) Expression applyDelegate(ForeachStatement fs, Expression flde,
```
Comment (WalterBright) [inline src/dmd/statementsem.d:1813]: Please do not create functions with no Ddoc comment explaining the parameters and returns.

File: src/dmd/statementsem.d [hunk 9]
```diff
@@ -1670,306 +1761,250 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
...
-                    // Easy case, a clean exit from the loop
-                    e = new CastExp(loc, e, Type.tvoid); // https://issues.dlang.org/show_bug.cgi?id=13899
-                    s = new ExpStatement(loc, e);
-                }
-                else
-                {
-                    // Construct a switch statement around the return value
-                    // of the apply function.
-                    auto a = new Statements();
+    private static extern(D) Expression applyArray(ForeachStatement fs, Expression flde,
```
Comment (WalterBright) [inline src/dmd/statementsem.d:1838]: Creating new functions with no Ddoc is not acceptable.

**Comment by WalterBright** [comment]
> This is a large PR, please review by commit.

What is the purpose of this PR?

**Comment by thewilsonator** [comment]
> Do them as smaller PRs.

Look at the commits.

File: src/dmd/statementsem.d (hunk 8 above)
Comment (thewilsonator) [reply src/dmd/statementsem.d:1813]: These are private functions

File: src/dmd/statementsem.d (hunk 9 above)
Comment (thewilsonator) [reply src/dmd/statementsem.d:1838]: ditto

**Comment by thewilsonator** [comment]
> What is the purpose of this PR?

I'm trying to get foreach semantic into a comprehensible state so that I can figure out why I can' t get the delegate to opApply to have a shared context. see https://github.com/dlang/dmd/pull/12290

**Comment by WalterBright** [comment]
Giant PRs consisting of independent commits is not acceptable. For example, if it produces a bug, I have to revert it as a group.

**Comment by WalterBright** [comment]
> I'm trying to get foreach semantic into a comprehensible state so that I can figure out why I can' t get the delegate to opApply to have a shared context. see #12290

"into a comprehensible state" is not a helpful rationale. I appreciate that you're trying to understand it, but upending the master repository upends other people working on it. You can modify it all you need to to understand it in your own fork (I often do this myself).

**Comment by WalterBright** [comment]
Each of the commits should be judged on its own merits, not as an all-or-nothing aggregate.

**Comment by Geod24** [comment]
> Giant PRs consisting of independent commits is not acceptable. For example, if it produces a bug, I have to revert it as a group.

You can `git revert <commit>` on the command line and revert whichever commit you want.

**Comment by Imperatorn** [comment]
I like this PR in general, but some constructive feedback:

The PR description could have been clearer to reduce the amount of questions, like 

"this refactoring is done because of x and enables y because reason a, b and c" 

or similar.

**Comment by WalterBright** [comment]
> You can git revert <commit> on the command line and revert whichever commit you want.

Or you can have independent commits in independent PRs, where the "revert" button works, where the test suite checks each one independently, where each can be reviewed on its own merits, where it's easier to figure out which caused a regression.

There is no advantage whatsoever in aggregating them.

**Pull request merged**